README.md
# bitsdojo_window

A Flutter package that makes it easy to customize and work with your Flutter desktop app window. 

Watch the tutorial to get started. Click the image below to watch the video: 

[![IMAGE ALT TEXT](https://img.youtube.com/vi/bee2AHQpGK4/0.jpg)](http://www.youtube.com/watch?v=bee2AHQpGK4 "Click to open")

<img src="https://github.com/bitsdojo/bitsdojo_window/blob/master/resources/screenshot.png">

**Features**:

    - Custom window frame - remove standard Windows titlebar and buttons
    - Hide window on startup
    - Show/hide window
    - Move window using Flutter widget
    - Minimize/Maximize/Restore/Close window
    - Set window size, minimum size and maximum size
    - Set window position
    - Set window alignment on screen (center/topLeft/topRight/bottomLeft/bottomRight)
    - Set window title


Currently working with Flutter desktop apps for **Windows**. macOS support is also planned in the future.

## Getting Started

Install the package using `pubspec.yaml`

Inside your application folder, go to `windows\runner\main.cpp` and add these two lines at the beginning of the file:

```
#include <bitsdojo_window/bitsdojo_window_plugin.h>
auto bdw = bitsdojo_window_configure(BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP);
```

If you don't want to use a custom frame and prefer the standard Windows titlebar and buttons, you can remove the `BDW_CUSTOM_FRAME` flag from the code above.

If you don't want to hide the window on startup, you can remove the `BDW_HIDE_ON_STARTUP` flag from the code above.

Now go to `lib\main.dart` and add this code in the `main` function right after `runApp(MyApp());` :

```
void main() {
  runApp(MyApp());

  // Add this code below

  doWhenWindowReady(() {
    var initialSize = Size(600, 450);
    appWindow.minSize = initialSize;
    appWindow.size = initialSize;
    appWindow.alignment = Alignment.center;
    appWindow.show();
  });
}
```
This will set an initial size and a minimum size for your application window, center it on the screen and show it on the screen.

You can find examples in the `example` folder.

Here is an example that displays this window:


```
import 'package:flutter/material.dart';
import 'package:bitsdojo_window/bitsdojo_window.dart';

void main() {
  runApp(MyApp());
  doWhenWindowReady(() {
    var win = appWindow;
    var initialSize = Size(600, 450);
    win.minSize = initialSize;
    win.size = initialSize;
    win.alignment = Alignment.center;
    win.title = "Custom window with Flutter";
    win.show();
  });
}

var borderColor = Color(0xFF805306);

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        home: Scaffold(
            body: WindowBorder(
                color: borderColor,
                width: 1,
                child: Row(children: [LeftSide(), RightSide()]))));
  }
}

var sidebarColor = Color(0xFFF6A00C);

class LeftSide extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
        width: 200,
        child: Container(
            color: sidebarColor,
            child: Column(
              children: [
                WindowTitleBarBox(child: MoveWindow()),
                Expanded(child: Container())
              ],
            )));
  }
}

var backgroundStartColor = Color(0xFFFFD500);
var backgroundEndColor = Color(0xFFF6A00C);

class RightSide extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Expanded(
        child: Container(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  colors: [backgroundStartColor, backgroundEndColor],
                  stops: [0.0, 1.0]),
            ),
            child: Column(children: [
              WindowTitleBarBox(
                  child: Row(children: [
                Expanded(child: MoveWindow()),
                WindowButtons()
              ])),
            ])));
  }
}

var buttonColors = WindowButtonColors(
    iconNormal: Color(0xFF805306),
    mouseOver: Color(0xFFF6A00C),
    mouseDown: Color(0xFF805306),
    iconMouseOver: Color(0xFF805306),
    iconMouseDown: Color(0xFFFFD500));

var closeButtonColors = WindowButtonColors(
    mouseOver: Colors.red[700],
    mouseDown: Colors.red[900],
    iconNormal: Color(0xFF805306),
    iconMouseOver: Colors.white);

class WindowButtons extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        MinimizeWindowButton(colors: buttonColors),
        MaximizeWindowButton(colors: buttonColors),
        CloseWindowButton(colors: closeButtonColors),
      ],
    );
```

TODO: More docs coming soon


CHANGELOG.md
## 0.0.4
    - Better integration with other plugins
## 0.0.3
    - Using dpi-aware values for title bar and buttons dimensions
    - Dynamically calculating default button padding instead of fixed one
## 0.0.2
    - Added video tutorial link
## 0.0.1

* Initial release
    - Custom window frame - remove standard Windows titlebar and buttons
    - Hide window on startup
    - Show/hide window
    - Minimize/Maximize/Restore/Close window
    - Move window using Flutter widget
    - Set window size, minimum size and maximum size
    - Set window position
    - Set window alignment on screen (center/topLeft/topRight/bottomLeft/bottomRight)
    - Set window title


example/switch.dart
/*
Don't forget to add these 2 lines at the beggining of windows\runner\main.cpp

#include <bitsdojo_window/bitsdojo_window_plugin.h>
auto bdw = bitsdojo_window_configure(BDW_CUSTOM_FRAME | BDW_HIDE_ON_STARTUP);

*/

import 'package:flutter/material.dart';
import 'package:bitsdojo_window/bitsdojo_window.dart';

class AppSkin {
  Color sidebar;
  Color backgroundStart;
  Color backgroundEnd;
  Color border;
  Color buttonMouseOver;
  Color buttonMouseDown;
  Color icon;
  Color iconMouseOver;
  Color iconMouseDown;

  AppSkin(
      {@required this.sidebar,
      @required this.backgroundStart,
      @required this.backgroundEnd,
      @required this.border,
      @required this.buttonMouseOver,
      @required this.buttonMouseDown,
      @required this.icon,
      @required this.iconMouseOver,
      @required this.iconMouseDown});
}

AppSkin yellowSkin = AppSkin(
  sidebar: Color(0xFFF6A00C),
  backgroundStart: Color(0xFFFFD500),
  backgroundEnd: Color(0xFFF6A00C),
  border: Color(0xFF805306),
  buttonMouseOver: Color(0xFFF6A00C),
  buttonMouseDown: Color(0xFF805306),
  icon: Color(0xFF805306),
  iconMouseOver: Color(0xFF805306),
  iconMouseDown: Color(0xFFFFD500),
);

AppSkin greenSkin = AppSkin(
    sidebar: Color(0xFF198A00),
    backgroundStart: Color(0xFF25C901),
    backgroundEnd: Color(0xFF198A00),
    border: Color(0xFF0C4500),
    buttonMouseOver: Color(0xFF198A00),
    buttonMouseDown: Color(0xFF0C4500),
    icon: Color(0xFF0C4500),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin blueSkin = AppSkin(
    sidebar: Color(0xFF125CDB),
    backgroundStart: Color(0xFF079BF2),
    backgroundEnd: Color(0xFF125CDB),
    border: Color(0xFF092E6E),
    buttonMouseOver: Color(0xFF125CDB),
    buttonMouseDown: Color(0xFF092E6E),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin purpleSkin = AppSkin(
    sidebar: Color(0xFF8700B2),
    backgroundStart: Color(0xFFCC00C5),
    backgroundEnd: Color(0xFF9A00CC),
    border: Color(0xFF730099),
    buttonMouseOver: Color(0xFF9A00CC),
    buttonMouseDown: Color(0xFF730099),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

AppSkin cherrySkin = AppSkin(
    sidebar: Color(0xFF850250),
    backgroundStart: Color(0xFFC90078),
    backgroundEnd: Color(0xFF850250),
    border: Color(0xFF700043),
    buttonMouseOver: Color(0xFF850250),
    buttonMouseDown: Color(0xFF700043),
    icon: Color(0xFFFFFFFF),
    iconMouseOver: Color(0xFFFFFFFF),
    iconMouseDown: Color(0xFFFFFFFF));

var skins = [yellowSkin, greenSkin, blueSkin, purpleSkin, cherrySkin];

class AppColors extends InheritedWidget {
  final AppSkin colors;
  final Widget child;
  const AppColors({
    Key key,
    @required this.colors,
    @required this.child,
  }) : super(key: key, child: child);

  static AppColors of(BuildContext context) {
    return context.dependOnInheritedWidgetOfExactType<AppColors>();
  }

  @override
  bool updateShouldNotify(AppColors old) => colors != old.colors;
}

void main() {
  runApp(MyApp());
  doWhenWindowReady(() {
    var initialSize = Size(600, 450);
    var win = appWindow;
    win.alignment = Alignment.center;
    win.minSize = initialSize;
    win.size = initialSize;
    win.title = "Switch colors Flutter app";
    win.show();
  });
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false, home: Scaffold(body: AppBody()));
  }
}

class AppBody extends StatefulWidget {
  @override
  _AppBodyState createState() => _AppBodyState();
}

class _AppBodyState extends State<AppBody> {
  int _skinNumber = 0;
  AppSkin _skin = skins[0];

  void switchSkin() {
    _skinNumber++;
    _skinNumber = (_skinNumber % skins.length);
    setState(() {
      _skin = skins[_skinNumber];
    });
  }

  @override
  Widget build(BuildContext context) {
    return AppColors(
        colors: _skin,
        child: WindowBorder(
            color: _skin.border,
            width: 1,
            child: Row(children: [
              LeftSide(),
              RightSide(
                onButtonPressed: switchSkin,
              )
            ])));
  }
}

class LeftSide extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    var colors = AppColors.of(context).colors;
    return SizedBox(
        width: 200,
        child: Container(
            color: colors.sidebar,
            child: Column(
              children: [WindowTitleBarBox(child: MoveWindow())],
            )));
  }
}

class RightSide extends StatelessWidget {
  final VoidCallback onButtonPressed;
  RightSide({Key key, this.onButtonPressed}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    var colors = AppColors.of(context).colors;
    return Expanded(
        child: Container(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  colors: [colors.backgroundStart, colors.backgroundEnd],
                  stops: [0.0, 1.0]),
            ),
            child: Column(children: [
              RightSideTopArea(),
              Expanded(
                child: Center(
                    child: RoundedFlatButton(
                        text: 'Switch', onPressed: () => onButtonPressed())),
              )
            ])));
  }
}

class RightSideTopArea extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return WindowTitleBarBox(
        child: Row(children: [Expanded(child: MoveWindow()), WindowButtons()]));
  }
}

class WindowButtons extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    var colors = AppColors.of(context).colors;
    var buttonColors = WindowButtonColors(
        iconNormal: colors.icon,
        iconMouseDown: colors.iconMouseDown,
        iconMouseOver: colors.iconMouseOver,
        mouseDown: colors.buttonMouseDown,
        mouseOver: colors.buttonMouseOver);
    var closeButtonColors = WindowButtonColors(
        mouseOver: Colors.red[700],
        mouseDown: Colors.red[900],
        iconNormal: colors.icon,
        iconMouseOver: Colors.white);

    return Row(
      children: [
        MinimizeWindowButton(colors: buttonColors),
        MaximizeWindowButton(colors: buttonColors),
        CloseWindowButton(colors: closeButtonColors),
      ],
    );
  }
}

class RoundedFlatButton extends StatelessWidget {
  final void Function() onPressed;
  final Color color;
  final Color mouseOverColor;
  final Color textColor;
  final String text;

  RoundedFlatButton(
      {Key key,
      this.onPressed,
      this.color,
      this.mouseOverColor,
      this.textColor,
      this.text})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return FlatButton(
      hoverColor: mouseOverColor ?? Colors.grey[800],
      color: color ?? Colors.grey[900],
      child: Padding(
        padding: const EdgeInsets.only(
          left: 10.0,
          right: 10.0,
        ),
        child: Text(text ?? 'Button',
            style: TextStyle(
                color: textColor ?? Colors.white,
                fontSize: 12.0,
                fontWeight: FontWeight.w600)),
      ),
      shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.circular(10.0),
          side: BorderSide(color: Colors.white)),
      onPressed: onPressed,
    );
  }
}
